### PR TITLE
[ENH] Speedup EAgglo by factor 5-10x

### PR DESCRIPTION
**Pull request by patrickzib**
### EAgglo
This PR improves the runtime of EAgglo by a factor of 5-10, by use of:

- NUMBA
- Reimplement Euclidean Distance (instead of scipy cdist)
- Reworked initial distance computations, if `self._member is None` (no groupings + Pandas needed)

### Experiments:

```python
from sktime.datasets import load_gun_point_segmentation
ts, period_size, cps = load_gun_point_segmentation()

# compute a segmentation
model = EAgglo(penalty=lambda x: np.mean(np.diff(np.sort(ts))))
fitted_model = model._fit(pd.DataFrame(ts[:2000]))
```

### After:  12 seconds

### CODE REVIEW

File: sktime/annotation/tests/test_eagglo.py [hunk 1]
```diff
@@ -126,3 +126,16 @@ def test_custom_penalty():
 
     assert np.allclose(cluster_actual, cluster_expected)
     assert np.allclose(fit_actual, fit_expected)
+
+
+# def test_eagglo_large():
```
Comment (lmmentel) [inline sktime/annotation/tests/test_eagglo.py:131]: Is the commented test needed?

**Review by lmmentel — COMMENTED** [review]
Left a small comment but all looks good!

**Comment by TonyBagnall** [comment]
superb, but could this use our numba Euclidean distance in distances rather than reimplement? @chrisholder

File: sktime/annotation/tests/test_eagglo.py (hunk 1 above)
Comment (patrickzib) [reply sktime/annotation/tests/test_eagglo.py:131]: Thanks. Only for benchmarking. Can remove it, once we finalize the PR :)

**Comment by patrickzib** [comment]
> superb, but could this use our numba Euclidean distance in distances rather than reimplement? @chrisholder

@TonyBagnall tbh: I do not understand how :)

The method I found is not numba`rized
> from sktime.distances._distance import euclidean_distance
> def euclidean_distance(x: np.ndarray, y: np.ndarray, **kwargs: Any) -> float:

Thus, I am not able to use it in a numba'rized function. 
I.e. to use it, I have to remove the decorators from the functions and the runtime increases to 35 seconds (from 15 seconds without).

File: sktime/annotation/tests/test_eagglo.py (hunk 1 above)
Comment (patrickzib) [reply sktime/annotation/tests/test_eagglo.py]: Removed :)

**Comment by TonyBagnall** [comment]
we need a distances notebook @chrisholder :) Will sort it tomorrow

**Review by lmmentel — APPROVED** [review]

**Pull request merged**